cmake/modules/FindGMock.cmake

```diff
@@ -113,172 +117,329 @@ function(GTEST_ADD_TESTS executable extra_args)
-    if("${GMOCK_VER}" STREQUAL "1.6.0" OR "${GMOCK_VER}" STREQUAL "1.7.0") 
-        ExternalProject_Add(
-            gtest
-            GIT_REPOSITORY "https://github.com/google/googletest.git"
-            GIT_TAG "release-${GMOCK_VER}"
-            PREFIX ${GMOCK_ROOT}
-            INSTALL_COMMAND ""
-            LOG_DOWNLOAD ON
-            LOG_CONFIGURE ON
-            LOG_BUILD ON
-            CMAKE_ARGS ${GTEST_CMAKE_ARGS}
-        )
-        ExternalProject_Add(
+if(NOT TARGET GTest::GTest)
+    add_library(GTest::GTest UNKNOWN IMPORTED)
+endif()
+if(NOT TARGET GTest::Main)
+    add_library(GTest::Main UNKNOWN IMPORTED)
+endif()
+
+if(NOT TARGET GMock::GMock)
+    add_library(GMock::GMock UNKNOWN IMPORTED)
+endif()
+
+if(NOT TARGET GMock::Main)
+    add_library(GMock::Main UNKNOWN IMPORTED)
+endif()
+
+set(GMOCK_LIBRARY_EXISTS OFF)
+set(GTEST_LIBRARY_EXISTS OFF)
+
+if(EXISTS "${GMOCK_LIBRARY}" OR EXISTS "${GMOCK_LIBRARY_DEBUG}" AND GMOCK_INCLUDE_DIR)
+    set(GMOCK_LIBRARY_EXISTS ON)
+endif()
+
+if(EXISTS "${GTEST_LIBRARY}" OR EXISTS "${GTEST_LIBRARY_DEBUG}" AND GTEST_INCLUDE_DIR)
+    set(GTEST_LIBRARY_EXISTS ON)
+endif()
+
+if(NOT (${GMOCK_LIBRARY_EXISTS} AND ${GTEST_LIBRARY_EXISTS}))
+
+    include(ExternalProject)
+
+    if(MSVC)
+        set(Suffix ".lib")
+    else()
+        set(Suffix ".a")
+    endif()
+
+    if(GTEST_USE_STATIC_LIBS)
+        set(GTEST_CMAKE_ARGS -Dgtest_force_shared_crt:BOOL=ON -DBUILD_SHARED_LIBS=OFF)
+        set(GTEST_LIBRARY_PREFIX ${CMAKE_STATIC_LIBRARY_PREFIX})
+    else()
+        set(GTEST_CMAKE_ARGS -DBUILD_SHARED_LIBS=ON)
+        set(GTEST_LIBRARY_PREFIX ${CMAKE_SHARED_LIBRARY_PREFIX})
+    endif()
+
+    if("${GMOCK_SRC_DIR}" STREQUAL "")
+        message(STATUS "Downloading GMock / GTest version ${GMOCK_VER} from git")
+        if("${GMOCK_VER}" STREQUAL "1.6.0" OR "${GMOCK_VER}" STREQUAL "1.7.0")
+            externalproject_add(
+                gtest
+                GIT_REPOSITORY "https://github.com/google/googletest.git"
+                GIT_TAG "release-${GMOCK_VER}"
+                PREFIX ${GMOCK_ROOT}
+                INSTALL_COMMAND ""
+                LOG_DOWNLOAD ON
+                LOG_CONFIGURE ON
+                LOG_BUILD ON
+                CMAKE_ARGS
+                    -Dgtest_disable_pthreads=${MINGW}
+                    ${GTEST_CMAKE_ARGS}
+            )
+            externalproject_add(
+                gmock
+                GIT_REPOSITORY "https://github.com/google/googlemock.git"
+                GIT_TAG "release-${GMOCK_VER}"
+                PREFIX ${GMOCK_ROOT}
+                INSTALL_COMMAND ""
+                LOG_DOWNLOAD ON
+                LOG_CONFIGURE ON
+                LOG_BUILD ON
+                CMAKE_ARGS
+                    -Dgtest_disable_pthreads=${MINGW}
+                    ${GTEST_CMAKE_ARGS}
+                DEPENDS gtest
+            )
+
+
+            add_dependencies(GTest::GTest gtest)
```

Comment:
If you specify `BUILD_BYPRODUCTS` you don't need these custom dependency declarations at all (as long as you use one of those byproducts in the library's `IMPORT_LOCATION`).